# Delinquency Telecom Model

# Definition:
Delinquency is a condition that arises when an activity or situation does not occur at its scheduled (or expected) date i.e., it occurs later than expected.
# Use Case: 
Many donors, experts, and microfinance institutions (MFI) have become convinced that using mobile financial services (MFS) is more convenient and efficient, and less costly, than the traditional high-touch model for delivering microfinance services. MFS becomes especially useful when targeting the unbanked poor living in remote areas. The implementation of MFS, though, has been uneven with both significant challenges and successes.
Today, microfinance is widely accepted as a poverty-reduction tool, representing $70 billion in outstanding loans and a global outreach of 200 million clients.

One of our Client in Telecom collaborates with an MFI to provide micro-credit on mobile balances to be paid back in 5 days. The Consumer is believed to be delinquent if he deviates from the path of paying back the loaned amount within 5 days.
The sample data from our client database is hereby given to you for the exercise.



# Exercise:
Create a delinquency model which can predict in terms of a probability for each loan transaction, whether the customer will be paying back the loaned amount within 5 days of insurance of loan 
(Label ‘1’ & ’0’)
Find Enclosed the Data Description File and The Sample Data for the Modeling Exercise.


# Loading the Data

In [17]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sn
import warnings
warnings.filterwarnings('ignore')

# Data Description
The following is the description of each of the columns of the data provided. We will first remove the columns that won't be useful in training the model.

In [18]:
pd.options.display.max_colwidth = 200
pd.read_excel('Data_Description.xlsx', sheet_name='Description')[['Variable', 'Definition']]

Variable  \
0                  label   
1                 msisdn   
2                    aon   
3           daily_decr30   
4           daily_decr90   
5               rental30   
6               rental90   
7      last_rech_date_ma   
8      last_rech_date_da   
9       last_rech_amt_ma   
10         cnt_ma_rech30   
11          fr_ma_rech30   
12     sumamnt_ma_rech30   
13  medianamnt_ma_rech30   
14  medianmarechprebal30   
15         cnt_ma_rech90   
16          fr_ma_rech90   
17     sumamnt_ma_rech90   
18  medianamnt_ma_rech90   
19  medianmarechprebal90   
20         cnt_da_rech30   
21          fr_da_rech30   
22         cnt_da_rech90   
23          fr_da_rech90   
24           cnt_loans30   
25          amnt_loans30   
26       maxamnt_loans30   
27    medianamnt_loans30   
28           cnt_loans90   
29          amnt_loans90   
30       maxamnt_loans90   
31    medianamnt_loans90   
32             payback30   
33             payback90   
34               pcircle   
35                 pdate   

                                                                                                              Definition  
0   Flag indicating whether the user paid back the credit amount within 5 days of issuing the loan{1:success, 0:failure}  
1                                                                                                  mobile number of user  
2                                                                                        age on cellular network in days  
3                                Daily amount spent from main account, averaged over last 30 days (in Indonesian Rupiah)  
4                                Daily amount spent from main account, averaged over last 90 days (in Indonesian Rupiah)  
5                                                                         Average main account balance over last 30 days  
6                                                                         Average main account balance over last 90 days  
7                                                                      Number of days till last recharge of main account  
8                                                                      Number of days till last recharge of data account  
9                                                         Amount of last recharge of main account (in Indonesian Rupiah)  
10                                                            Number of times main account got recharged in last 30 days  
11                                                                   Frequency of main account recharged in last 30 days  
12                                     Total amount of recharge in main account over last 30 days (in Indonesian Rupiah)  
13             Median of amount of recharges done in main account over last 30 days at user level (in Indonesian Rupiah)  
14              Median of main account balance just before recharge in last 30 days at user level (in Indonesian Rupiah)  
15                                                            Number of times main account got recharged in last 90 days  
16                                                                   Frequency of main account recharged in last 90 days  
17                                     Total amount of recharge in main account over last 90 days (in Indonesian Rupiah)  
18             Median of amount of recharges done in main account over last 90 days at user level (in Indonesian Rupiah)  
19              Median of main account balance just before recharge in last 90 days at user level (in Indonesian Rupiah)  
20                                                            Number of times data account got recharged in last 30 days  
21                                                                   Frequency of data account recharged in last 30 days  
22                                                            Number of times data account got recharged in last 90 days  
23                         

In [19]:
df = pd.read_csv('sample_data_intw.csv')
df.head()
df.shape

Unnamed: 0  label       msisdn    aon  daily_decr30  daily_decr90  \
0           1      0  21408I70789  272.0   3055.050000   3065.150000   
1           2      1  76462I70374  712.0  12122.000000  12124.750000   
2           3      1  17943I70372  535.0   1398.000000   1398.000000   
3           4      1  55773I70781  241.0     21.228000     21.228000   
4           5      1  03813I82730  947.0    150.619333    150.619333   

   rental30  rental90  last_rech_date_ma  last_rech_date_da     ...      \
0    220.13    260.13                2.0                0.0     ...       
1   3691.26   3691.26               20.0                0.0     ...       
2    900.13    900.13                3.0                0.0     ...       
3    159.42    159.42               41.0                0.0     ...       
4   1098.90   1098.90                4.0                0.0     ...       

   maxamnt_loans30  medianamnt_loans30  cnt_loans90  amnt_loans90  \
0              6.0                 0.0          2.0            12   
1             12.0                 0.0          1.0            12   
2              6.0                 0.0          1.0             6   
3              6.0                 0.0          2.0            12   
4              6.0                 0.0          7.0            42   

   maxamnt_loans90  medianamnt_loans90  payback30  payback90  pcircle  \
0                6                 0.0  29.000000  29.000000      UPW   
1               12                 0.0   0.000000   0.000000      UPW   
2                6                 0.0   0.000000   0.000000      UPW   
3                6                 0.0   0.000000   0.000000      UPW   
4                6                 0.0   2.333333   2.333333      UPW   

        pdate  
0  2016-07-20  
1  2016-08-10  
2  2016-08-19  
3  2016-06-06  
4  2016-06-22  

[5 rows x 37 columns]

(209593, 37)

It is always a good idea to first have a look at the data. `df.describe()` summarizes valuable statistics about the data.

It will also help us decide which columns we can eliminate straightaway.

In [20]:
df.describe()

Unnamed: 0          label            aon   daily_decr30  \
count  209593.000000  209593.000000  209593.000000  209593.000000   
mean   104797.000000       0.875177    8112.343445    5381.402289   
std     60504.431823       0.330519   75696.082531    9220.623400   
min         1.000000       0.000000     -48.000000     -93.012667   
25%     52399.000000       1.000000     246.000000      42.440000   
50%    104797.000000       1.000000     527.000000    1469.175667   
75%    157195.000000       1.000000     982.000000    7244.000000   
max    209593.000000       1.000000  999860.755168  265926.000000   

        daily_decr90       rental30       rental90  last_rech_date_ma  \
count  209593.000000  209593.000000  209593.000000      209593.000000   
mean     6082.515068    2692.581910    3483.406534        3755.847800   
std     10918.812767    4308.586781    5770.461279       53905.892230   
min       -93.012667  -23737.140000  -24720.580000         -29.000000   
25%        42.692000     280.420000     300.260000           1.000000   
50%      1500.000000    1083.570000    1334.000000           3.000000   
75%      7802.790000    3356.940000    4201.790000           7.000000   
max    320630.000000  198926.110000  200148.110000      998650.377733   

       last_rech_date_da  last_rech_amt_ma      ...          cnt_loans30  \
count      209593.000000     209593.000000      ...        209593.000000   
mean         3712.202921       2064.452797      ...             2.758981   
std         53374.833430       2370.786034      ...             2.554502   
min           -29.000000          0.000000      ...             0.000000   
25%             0.000000        770.000000      ...             1.000000   
50%             0.000000       1539.000000      ...             2.000000   
75%             0.000000       2309.000000      ...             4.000000   
max        999171.809410      55000.000000      ...            50.000000   

        amnt_loans30  maxamnt_loans30  medianamnt_loans30    cnt_loans90  \
count  209593.000000    209593.000000       209593.000000  209593.000000   
mean       17.952021       274.658747            0.054029      18.520919   
std        17.379741      4245.264648            0.218039     224.797423   
min         0.000000         0.000000            0.000000       0.000000   
25%         6.000000         6.000000            0.000000       1.000000   
50%        12.000000         6.000000            0.000000       2.000000   
75%        24.000000         6.000000            0.000000       5.000000   
max       306.000000     99864.560864            3.000000    4997.517944   

        amnt_loans90  maxamnt_loans90  medianamnt_loans90      payback30  \
count  209593.000000    209593.000000       209593.000000  209593.000000   
mean       23.645398         6.703134            0.046077       3.398826   
std        26.469861         2.103864            0.200692       8.813729   
min         0.000000         0.000000            0.000000       0.000000   
25%         6.000000         6.000000            0.000000       0.000000   
50%        12.000000         6.000000            0.000000       0.000000   
75%        30.000000         6.000000            0.000000       3.750000   
max       438.000000        12.000000            3.000000     171.500000   

           payback90  
count  209593.000000  
mean        4.321485  
std        10.308108  
min         0.000000  
25%         0.000000  
50%         1.666667  
75%         4.500000  
max       171.500000  

[8 rows x 34 columns]

As we can see below, the data is highly skewed. The majority of data items in the dataset represent items belonging to the `positive` class, i.e., those who have paid back the loan in time.

In [21]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(df, test_size=0.2,random_state=1)
print("training data is :", train.shape)
train['label'].value_counts()
print("Test data is:", test.shape)
test['label'].value_counts()

training data is : (167674, 37)


1    146692
0     20982
Name: label, dtype: int64

Test data is: (41919, 37)


1    36739
0     5180
Name: label, dtype: int64

# Preprocessing the data

First we will remove the columns that won't be fed to the model. All the entries need to be converted to numeric values first. Note that we have set `errors='coerce'` so that any non-numeric values are set to NaN. Then, check if there exist any NaN entries.

In [22]:
train.drop(['Unnamed: 0','msisdn','pcircle','pdate'],axis=1,inplace=True)
test.drop(['Unnamed: 0','msisdn','pcircle','pdate'],axis=1,inplace=True)

In [23]:
train.apply(lambda x: pd.to_numeric(x, errors='coerce'))
test.apply(lambda x: pd.to_numeric(x, errors='coerce'))

label            aon  daily_decr30  daily_decr90  rental30  rental90  \
69724       0     240.000000      8.000000      8.000000   1569.60   1569.60   
136917      1     742.000000     12.900000     12.900000     44.10     44.10   
172099      1      91.000000  19012.634667  19065.690000  14002.06  15832.84   
88803       1     466.000000   8468.000000   8734.440000    800.52   1040.52   
70250       1    1405.000000  17029.748000  17121.320000   2359.28   3365.48   
2932        0     697.000000  12056.191667  12065.230000    445.96    475.16   
207129      1     476.000000      3.803000      3.803000     74.37     74.37   
43610       0    1603.000000      0.013333      0.013333      0.00      0.00   
19283       1    1399.000000   7882.000000   9389.000000   2618.51   2678.51   
179502      1     927.000000  17757.000000  20740.420000  12031.79  21220.86   
180439      1    1260.000000      5.643333      5.643333    374.90    374.90   
122835      0     630.000000   4953.733333   4964.000000   3982.85   5981.11   
64018       0     151.000000      0.633333      0.633333      0.00      0.00   
191071      0     296.000000    500.195000    500.450000      0.00      0.00   
55740       0     967.000000   5167.866667   5183.500000   3485.36   3978.30   
102728      0     290.000000      6.933333      6.933333     25.61     25.61   
92293       1      63.000000      1.460667      1.460667     43.26     43.26   
119512      1    1332.000000  16934.733333  17027.500000   2924.78   3366.98   
49662       1    1136.000000   3890.780000   3903.670000   1704.04   1754.04   
197044      1     578.000000  12845.000000  16298.860000   2660.38   5689.77   
198462      0      51.000000      0.000000      0.000000    357.40    357.40   
158886      1     213.000000     15.288000     15.288000     56.00     56.00   
205066      1     372.000000     27.505333     27.505333   3668.28   3668.28   
186357      1    1395.000000   8179.332667   8251.090000    513.94    264.90   
110351      1    1389.000000   1403.815000   1407.630000    297.11    494.66   
130659      1    1054.000000    126.298667    126.298667   1947.96   1947.96   
81427       1     162.000000     89.505000     89.505000   -175.76   -175.76   
9896        1     384.000000   4110.400000   4126.000000    423.69    759.03   
21750       1     956.000000   2427.115333   2460.210000   2026.16   2260.96   
129112      1     912.000000   5964.400000   6004.000000   1636.33   1903.19   
...       ...            ...           ...           ...       ...       ...   
182112      1    1945.000000   8087.379000   8171.790000   4397.96   6252.05   
175582      1     279.000000   6031.500000   6052.500000    557.79    729.39   
26103       1     128.000000    143.520000    143.520000    548.16    548.16   
148209      1     102.000000   3963.000000   7751.310000    933.59   2932.37   
150505      1     205.000000  24998.000000  32994.880000   3922.94   5824.81   
31959       0     452.000000    800.000000    800.000000    711.65    847.24   
92740       1     142.000000    259.553333    259.553333  -1743.20  -1743.20   
188317      1     229.000000     36.233333     36.233333     80.00     80.00   
103068      1     785.000000   4050.000000   4050.000000    599.38   2034.83   
19946       0  892075.528158    500.000000    500.000000    660.98    695.78   
139516      1     861.000000   1507.148667   1518.710000    512.56    576.56   
195467      1     295.000000  37994.100000  38282.300000   4390.79   5357.39   
51668       1     590.000000   2517.755000   2545.731667  14000.42  16157.26   
166802      1     257.000000   3750.000000   3868.060000    399.25    786.25   
176485      1     387.000000     34.752667     34.752667    482.86    482.86   
21758       1     126.000000   2300.000000   2727.500000   1938.25   2472.59   
167302      1     128.000000      8.146667      8.146667    620.50    620.50   
31228       1     480.000000   7388.000000   7493.530000   2003.29   2120.29   
174829  

label            aon  daily_decr30  daily_decr90  rental30  rental90  \
195331      1     201.000000   2842.400000   2853.000000    439.25    474.89   
61584       1     189.000000     73.905000     73.905000    448.76    448.76   
103931      1     192.000000     21.237000     21.237000    182.10    182.10   
113088      1    1046.000000     87.583667     87.583667    -66.69    -66.69   
95108       1     183.000000  14969.120000  15164.900000   4321.07   5750.55   
47805       1      98.000000     41.600000     41.600000   1195.22   1195.22   
72268       1    1991.000000   7580.000000   8200.200000   5991.69   9656.22   
149926      1    1353.000000     14.842667     14.842667    159.28    159.28   
31946       1      87.000000     77.345000     77.345000   2586.30   2586.30   
129634      1     229.000000   1230.400000   1238.000000    218.84    248.66   
64927       1     467.000000    137.859333    137.859333   -347.82   -347.82   
108505      1    2258.000000   7450.870000   8617.120000   2847.52   3241.98   
40847       1    1853.000000   1460.000000   1460.000000    427.54    619.16   
125479      1     670.000000  19351.000000  26910.520000   5604.44   7574.15   
104400      1     638.000000   5624.900000   5669.000000    618.24    806.16   
206650      1     495.000000  16469.500000  16571.500000   3648.75   8067.39   
95374       1    1449.000000   5240.004667   5261.420000   2726.65   3434.37   
116943      1    1455.000000    114.890000    114.890000   2931.30   2931.30   
145595      1     662.000000     57.674667     57.674667   3213.28   3213.28   
191197      1    1479.000000   7785.522667   7954.710000   3119.44   4367.94   
16037       1     751.000000  23278.300667  23432.020000   3993.57   4885.61   
77919       1     339.000000   6788.040000   6908.240000   1240.20   2437.95   
56894       1     157.000000   5636.000000   5636.000000   2014.25   2509.85   
117864      1    1761.000000   4807.000000   9670.690000    398.24   2079.96   
200405      1     643.000000    934.281333    937.230000   1117.48   1197.30   
54804       1     530.000000   2685.120000   2703.800000    536.22    595.80   
207220      1    1142.000000     41.647667     41.647667    -91.28    -91.28   
176083      1     926.000000   2157.536000   2159.520000    144.71    149.70   
90492       1     354.000000      7.033333      7.033333     53.88     53.88   
22034       1    1616.000000  14938.500000  17630.870000  12304.32  15041.82   
...       ...            ...           ...           ...       ...       ...   
133707      1     426.000000     83.561333     83.561333    220.64    220.64   
157731      0     698.000000      0.000000      0.000000   1548.00   1548.00   
17171       1     164.000000     14.960000     14.960000    466.07    466.07   
155973      1     -46.000000     17.000000     17.000000     64.56     64.56   
110716      1     560.000000  13261.875000  13296.250000   1729.62   1879.62   
68401       1     283.000000  15717.570667  15763.140000   8109.08  11289.62   
35140       1     282.000000  14686.000000  14698.450000   2206.07   2206.07   
41682       1     558.000000   1200.000000   1200.000000    399.26    399.26   
72695       1    1134.000000     81.492000     81.492000    -39.78    -39.78   
39287       0     201.000000     20.863333     20.863333   5942.20   5942.20   
82602       1     253.000000    712.800000    724.000000     47.52     89.10   
152986      0  653800.392174     43.333333     43.333333   6692.28   7154.84   
176378      1     467.000000     27.086667     27.086667     53.65     53.65   
11407       1     392.000000  15669.000000  21600.170000   5095.75   5137.89   
46895       1    1495.000000    105.357000    105.357000   1038.33   1038.33   
6287        1     934.000000   3433.000000   3433.000000    200.13    200.13   
52343       1     223.000000    909.533333    911.000000     50.44     58.20   
35310       0    1006.000000    700.000000    700.000000    125.28    313.20   
207925  

In [24]:
train.isnull().any()
test.isnull().any()

label                   False
aon                     False
daily_decr30            False
daily_decr90            False
rental30                False
rental90                False
last_rech_date_ma       False
last_rech_date_da       False
last_rech_amt_ma        False
cnt_ma_rech30           False
fr_ma_rech30            False
sumamnt_ma_rech30       False
medianamnt_ma_rech30    False
medianmarechprebal30    False
cnt_ma_rech90           False
fr_ma_rech90            False
sumamnt_ma_rech90       False
medianamnt_ma_rech90    False
medianmarechprebal90    False
cnt_da_rech30           False
fr_da_rech30            False
cnt_da_rech90           False
fr_da_rech90            False
cnt_loans30             False
amnt_loans30            False
maxamnt_loans30         False
medianamnt_loans30      False
cnt_loans90             False
amnt_loans90            False
maxamnt_loans90         False
medianamnt_loans90      False
payback30               False
payback90               False
dtype: boo

label                   False
aon                     False
daily_decr30            False
daily_decr90            False
rental30                False
rental90                False
last_rech_date_ma       False
last_rech_date_da       False
last_rech_amt_ma        False
cnt_ma_rech30           False
fr_ma_rech30            False
sumamnt_ma_rech30       False
medianamnt_ma_rech30    False
medianmarechprebal30    False
cnt_ma_rech90           False
fr_ma_rech90            False
sumamnt_ma_rech90       False
medianamnt_ma_rech90    False
medianmarechprebal90    False
cnt_da_rech30           False
fr_da_rech30            False
cnt_da_rech90           False
fr_da_rech90            False
cnt_loans30             False
amnt_loans30            False
maxamnt_loans30         False
medianamnt_loans30      False
cnt_loans90             False
amnt_loans90            False
maxamnt_loans90         False
medianamnt_loans90      False
payback30               False
payback90               False
dtype: boo

There are no null entries in the data, so we are good to go! Otherwise, we can fill those entries by their respective column's mean by uncommenting the following cell

In [25]:
# train.fillna(train.mean(), inplace=True)
# test.fillna(test.mean(), inplace=True)

When the majority of data items in our dataset represents items belonging to one class, we say the dataset is skewed or imbalanced. To prevent the model from overfitting we will try resampling and data augmentation separately and see which technique fares better.

### Resampling: Oversampling minority class
Over-sampling involves adding more examples from the minority class to the training dataset so that the model does not overfit on the majority class. Here we are resampling such that the new no. of negatives now equal the old no. of posiitves. The simplest implementation of over-sampling is to duplicate random records from the minority class.

In [26]:
from sklearn.utils import resample
from scipy import stats

In [27]:
def oversample_minority(x_train, y_train) :
    train_data = x_train
    train_data['label'] = y_train
    
    train_positive=train_data[train_data.label==1]
    train_negative=train[train_data.label==0]
    
    print("\nApplying oversampling on training data...")
    
    train_upsamp=resample(train_negative,replace=True,n_samples=train_positive.shape[0],random_state=1)
    train_upsamp=pd.concat([train_positive,train_upsamp])
    
    print("\nupsampled train:", train_upsamp.shape)
    print(train_upsamp['label'].value_counts())
    
    y_train = train_upsamp['label']
    x_train = train_upsamp.drop(['label'], axis=1)
    
    return x_train, y_train

print("\ntrain:", train.shape)
print(train['label'].value_counts())
    


train: (167674, 33)
1    146692
0     20982
Name: label, dtype: int64


In [38]:
!pip install msgpack

You are using pip version 10.0.1, however version 20.2b1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


### Resampling: Undersampling majority class
Under-sampling involves removing samples from the majority class. Although, we are losing some information, however, it is still worth giving a shot.

In [28]:
def downsample_minority(x_train, y_train) :
    train_data = x_train
    train_data['label'] = y_train
    
    train_positive=train_data[train_data.label==1]
    train_negative=train[train_data.label==0]
    
    
    print("\nApplying downsampling on training data...")
    
    train_downsamp=resample(train_positive,replace=False,n_samples=5*train_negative.shape[0],random_state=1)
    train_downsamp=pd.concat([train_negative,train_downsamp])
    
    print("\ndownsampled train:", train_downsamp.shape)
    print(train_downsamp['label'].value_counts())
    
    y_train = train_downsamp['label']
    x_train = train_downsamp.drop(['label'], axis=1)
    
    return x_train, y_train

print("\ntrain:", train.shape)
print(train['label'].value_counts())


train: (167674, 33)
1    146692
0     20982
Name: label, dtype: int64


### Resampling: SMOTE

Synthetic Minority Oversampling Technique (SMOTE) is a data augmentation technique that creates synthetic minority class examples. It used when there is a class imbalance in the training data.

In [40]:
!pip3 install imbalanced-learn

You are using pip version 9.0.1, however version 20.1.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [37]:
!pip install -U imbalanced-learn

  Using cached https://files.pythonhosted.org/packages/c8/81/8db4d87b03b998fda7c6f835d807c9ae4e3b141f978597b8d7f31600be15/imbalanced_learn-0.7.0-py3-none-any.whl
Requirement not upgraded as not directly required: scipy>=0.19.1 in c:\users\mratu\anaconda3\lib\site-packages (from imbalanced-learn) (1.1.0)
Requirement not upgraded as not directly required: numpy>=1.13.3 in c:\users\mratu\anaconda3\lib\site-packages (from imbalanced-learn) (1.14.3)
Requirement not upgraded as not directly required: joblib>=0.11 in c:\users\mratu\anaconda3\lib\site-packages (from imbalanced-learn) (0.16.0)
Requirement not upgraded as not directly required: scikit-learn>=0.23 in c:\users\mratu\anaconda3\lib\site-packages (from imbalanced-learn) (0.23.1)
Requirement not upgraded as not directly required: threadpoolctl>=2.0.0 in c:\users\mratu\anaconda3\lib\site-packages (from scikit-learn>=0.23->imbalanced-learn) (2.1.0)


distributed 1.21.8 requires msgpack, which is not installed.
You are using pip version 10.0.1, however version 20.2b1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [14]:
def smote_resampling(x_train, y_train) :
    
    print("\nTrain:", x_train.shape)
    print(y_train.value_counts())
    print("\nApplying SMOTE...")
    
    sm = SMOTE(random_state=27)
    x_train, y_train = sm.fit_sample(x_train, y_train)
    
    print("\nTrain:", x_train.shape)
    print(y_train.value_counts())
    
    return x_train, y_train

### Data Normalization
Standardization of datasets is a common requirement for many machine learning estimators.

Normalization is the process of scaling individual samples to have unit norm. We are going to use the Scikit-Learn's implementation of Normalization.

In [15]:
from sklearn.preprocessing import Normalizer

In [16]:
def normalize_data(x_train, x_test) :
    cols = [str(i) for i in x_train.columns]
    
    print("\nApplying data normalization...")
    train_normalized = Normalizer().fit_transform(x_train)
    train_normalized = pd.DataFrame(train_normalized, columns=cols)
    test_normalized = Normalizer().fit_transform(x_test)
    test_normalized = pd.DataFrame(test_normalized, columns=cols)
    
    x_train = train_normalized
    x_test = test_normalized
    
    return x_train, x_test

### Feature Selection with RFE

RFE stands for Recursive Feature Elimination. Finding optimal features for our model to train on is very important.

This technique begins by building a model on the entire set of predictors and computing an importance score for each predictor. The least important predictor(s) are then removed, the model is re-built, and importance scores are computed again.

Since RFE takes a lot of time to execute, I have already run RFE on training data with `num_features = 25` and `num_features=7`. 

`features_RFE_25` and `features_RFE_7` denote the selected 25 and 7 features respectively out of the 32 features that we have so far.

Feel free to uncomment the lines `fit=rfe.fit(x,y_train)` and `selected_features_boolean = fit.support_` to run RFE on specified `num_features`.

The function below will return `x_train` with only `num_features` no. of columns.


In [17]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import RFE


In [18]:
def feature_ranking(x_train, y_train, x_test, num_features) :
    
    print("\nApplying feature selection...")

    features_RFE_25 = [ True,  True,  True,  True,  True,  True, False,  True,  True,
                True,  True,  True,  True,  True,  True,  True,  True,  True,
               False, False, False, False,  True,  True,  True, False,  True,
                True,  True, False,  True,  True]
    
    features_RFE_7 = [False,  True,  True,  True, False,  True, False, False, False, False, False, False,
             False, False, False,  True, False, False, False, False, False, False, False, False,
              True, False, False, False,  True, False, False, False]
    
    rfe=RFE(RandomForestClassifier(verbose=1), num_features)
    x = x_train.values
    selected_features_boolean = features_RFE_7
    
#     fit=rfe.fit(x,y_train)
#     selected_features_boolean = fit.support_
    
    
    print("\nSelected features:", x_train.columns[selected_features_boolean])
    
    x_train = x_train[x_train.columns[selected_features_boolean]]
    x_test = x_test[x_test.columns[selected_features_boolean]]
    
    return x_train, x_test

# Training and Prediction

In [19]:
from sklearn.metrics import accuracy_score,precision_score,recall_score
from sklearn.metrics import classification_report

In [20]:
def train_and_predict(x_train, y_train, x_test, y_test, model) :
    print("\nTraining model...")
    model.fit(x_train, y_train)
    
    print("\nPredicting on test data...")
    prediction = model.predict(x_test)
    
    accuracy = round(accuracy_score(prediction,y_test)*100,2)
    precision = round(precision_score(prediction,y_test)*100,2)
    recall = round(recall_score(prediction,y_test)*100,2)
    
    print("\naccuracy:", accuracy)
    print("precision:", precision)
    print("recall:", recall)
    
    print(classification_report(prediction,y_test))

### Training and Prediction
 - Model: Random Forest Classifier
 - Resampling: SMOTE
 - Feature selection: Recursive Feature Extraction (RFE)

In [21]:
y_train = train['label']
y_test = test['label']

x_train = train.drop(['label'], axis=1)
x_test = test.drop(['label'], axis=1)

x_train, y_train = smote_resampling(x_train, y_train)

model1 = RandomForestClassifier(verbose=1)

x_train, x_test = normalize_data(x_train, x_test)

x_train, x_test = feature_ranking(x_train, y_train, x_test, 7)



Train: (167674, 32)
1    146692
0     20982
Name: label, dtype: int64

Applying SMOTE...

Train: (293384, 32)
1    146692
0    146692
Name: label, dtype: int64

Applying data normalization...

Applying feature selection...

Selected features: Index(['daily_decr30', 'daily_decr90', 'rental30', 'last_rech_date_ma',
       'sumamnt_ma_rech90', 'maxamnt_loans30', 'maxamnt_loans90'],
      dtype='object')


In [22]:
train_and_predict(x_train, y_train, x_test, y_test, model1)


Training model...


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:  1.1min finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.



Predicting on test data...

accuracy: 85.72
precision: 88.36
recall: 95.01
              precision    recall  f1-score   support

           0       0.67      0.45      0.54      7752
           1       0.88      0.95      0.92     34167

    accuracy                           0.86     41919
   macro avg       0.78      0.70      0.73     41919
weighted avg       0.84      0.86      0.85     41919



[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    1.0s finished


### Training and Prediction
 - Model: Random Forest Classifier
 - Resampling: Oversampling minority
 - Feature selection: Recursive Feature Extraction (RFE)

In [23]:
y_train = train['label']
y_test = test['label']

x_train = train.drop(['label'], axis=1)
x_test = test.drop(['label'], axis=1)

x_train, y_train = oversample_minority(x_train, y_train)

x_train, x_test = normalize_data(x_train, x_test)

model2 = RandomForestClassifier(verbose=1)

x_train, x_test = feature_ranking(x_train, y_train, x_test, 25)


train: (167674, 33)
1    146692
0     20982
Name: label, dtype: int64

Applying oversampling on training data...

upsampled train: (293384, 33)
1    146692
0    146692
Name: label, dtype: int64

Applying data normalization...

Applying feature selection...

Selected features: Index(['daily_decr30', 'daily_decr90', 'rental30', 'last_rech_date_ma',
       'sumamnt_ma_rech90', 'maxamnt_loans30', 'maxamnt_loans90'],
      dtype='object')


In [24]:
train_and_predict(x_train, y_train, x_test, y_test, model2)


Training model...


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:   56.5s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.



Predicting on test data...

accuracy: 90.29
precision: 96.07
recall: 93.07
              precision    recall  f1-score   support

           0       0.49      0.64      0.56      3994
           1       0.96      0.93      0.95     37925

    accuracy                           0.90     41919
   macro avg       0.73      0.78      0.75     41919
weighted avg       0.92      0.90      0.91     41919



[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.8s finished


### Training and Prediction
 - Model: Random Forest Classifier
 - Resampling: Downsampling majority
 - Feature selection: Recursive Feature Extraction (RFE)

In [25]:
y_train = train['label']
y_test = test['label']

x_train = train.drop(['label'], axis=1)
x_test = test.drop(['label'], axis=1)

x_train, y_train = downsample_minority(x_train, y_train)

x_train, x_test = normalize_data(x_train, x_test)

model3 = RandomForestClassifier(verbose=1)

x_train, x_test = feature_ranking(x_train, y_train, x_test, 25)


train: (167674, 33)
1    146692
0     20982
Name: label, dtype: int64

Applying downsampling on training data...

downsampled train: (125892, 33)
1    104910
0     20982
Name: label, dtype: int64

Applying data normalization...

Applying feature selection...

Selected features: Index(['daily_decr30', 'daily_decr90', 'rental30', 'last_rech_date_ma',
       'sumamnt_ma_rech90', 'maxamnt_loans30', 'maxamnt_loans90'],
      dtype='object')


In [26]:
train_and_predict(x_train, y_train, x_test, y_test, model3)


Training model...


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:   29.5s finished
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.



Predicting on test data...

accuracy: 90.58
precision: 96.55
recall: 92.98
              precision    recall  f1-score   support

           0       0.48      0.66      0.56      3768
           1       0.97      0.93      0.95     38151

    accuracy                           0.91     41919
   macro avg       0.72      0.80      0.75     41919
weighted avg       0.92      0.91      0.91     41919



[Parallel(n_jobs=1)]: Done 100 out of 100 | elapsed:    0.8s finished


### Training and Prediction
 - Model: XGBoost
 - Resampling: SMOTE
 - Feature selection: Recursive Feature Extraction (RFE)

In [27]:
from xgboost import XGBClassifier
y_train = train['label']
y_test = test['label']

x_train = train.drop(['label'], axis=1)
x_test = test.drop(['label'], axis=1)

x_train, y_train = smote_resampling(x_train, y_train)

x_train, x_test = normalize_data(x_train, x_test)

model4 = XGBClassifier(verbose=1)

x_train, x_test = feature_ranking(x_train, y_train, x_test, 25)


Train: (167674, 32)
1    146692
0     20982
Name: label, dtype: int64

Applying SMOTE...

Train: (293384, 32)
1    146692
0    146692
Name: label, dtype: int64

Applying data normalization...

Applying feature selection...

Selected features: Index(['daily_decr30', 'daily_decr90', 'rental30', 'last_rech_date_ma',
       'sumamnt_ma_rech90', 'maxamnt_loans30', 'maxamnt_loans90'],
      dtype='object')


In [28]:
train_and_predict(x_train, y_train, x_test, y_test, model4)


Training model...

Predicting on test data...

accuracy: 81.99
precision: 82.68
recall: 96.24
              precision    recall  f1-score   support

           0       0.77      0.39      0.51     10355
           1       0.83      0.96      0.89     31564

    accuracy                           0.82     41919
   macro avg       0.80      0.67      0.70     41919
weighted avg       0.81      0.82      0.80     41919



In [43]:
!pip install  PrettyTable

You are using pip version 10.0.1, however version 20.2b1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [44]:
from prettytable import PrettyTable
    
x = PrettyTable()

x.field_names = ["Model","Resampling","Feature Selection", "Accuracy"]
x.add_row(["Random Forest Classifier","Smote","Recursive feature Extraction",0.86])
x.add_row(["Random Forest Classifier","Oversampling minority","Recursive feature Extraction",0.90])
x.add_row(["Random Forest Classifier","Downsampling majority","Recursive feature Extraction",0.91 ])
x.add_row(["XGBoost","SMOTE","Recursive feature Extraction",0.82])



print(x)

+--------------------------+-----------------------+------------------------------+----------+
|          Model           |       Resampling      |      Feature Selection       | Accuracy |
+--------------------------+-----------------------+------------------------------+----------+
| Random Forest Classifier |         Smote         | Recursive feature Extraction |   0.86   |
| Random Forest Classifier | Oversampling minority | Recursive feature Extraction |   0.9    |
| Random Forest Classifier | Downsampling majority | Recursive feature Extraction |   0.91   |
|         XGBoost          |         SMOTE         | Recursive feature Extraction |   0.82   |
+--------------------------+-----------------------+------------------------------+----------+


# Conclusion
 - Used various permutations of the following models
     - Random Forest Classifier
     - XGBoost
 - The training data was highly skewed, so tried various resampling methods
     - Oversampling of majority class
     - Downsampling of minority class
     - SMOTE
 - Also used Recursive Feature Elimination (RFE) to extract only the relavant features to train the model on
 - Among the models the one with the highest accuracy (90.58%) was the following
     - Random Forest Classifier, Downsampling Minority, RFE with 7 features extracted
 - Although, upon analyzing the complete classification report we find that the model doesn't generalize well to the minority class.
 - The recall values for all the models for the `negative (false/0)` case is pretty low.
 - This clearly indicates there is a need to improve the performance of the model
 